xt_training/utils/logging.py

```diff
@@ -37,4 +40,41 @@ def flush(self):
         self.stdout.flush()
 
     def __del__(self):
-        self.close()
\ No newline at end of file
+        self.close()
+
+
+PATCH_HEADER = '''
+To replicate the git state used for this checkpoint, run the following:
+
+    $ git checkout <commit hash shown below>
+    $ git apply <path to this file>
+'''
+
+
+def _save_state(save_dir, config_path):
+    # Save config file
+    try:
+        if isinstance(config_path, str):
+            shutil.copy(config_path, f'{save_dir}/config.py')
+        else:
+            shutil.copy(config_path['__file__'], f'{save_dir}/config.py')
```

Comment:
Very minor but gave you looked into using pathlib.Path for managing file paths? I find it preferable because it handles the string formatting for us. So in this can `f'{save_dir}/config.py'` would become `save_dir.joinpath('config.py)` (assuming save_dir is a pathlib.Path object).

Reply:
I agree that `pathlib.Path` is superior. However, I won't use that here for consistency with the rest of the package. Though we should probably make a separate PR that changes that over everywhere. What I will do right now is change it to use `os.path.join` as that is both much safer and more consistent with the rest of the package.

I will add something to the backlog for moving over to pathlib.Path